internal/controllers/manager.go

```diff
@@ -0,0 +1,84 @@
+package controllers
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/rest"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+
+	"github.com/tilt-dev/tilt/internal/hud/server"
+	"github.com/tilt-dev/tilt/internal/store"
+	corev1alpha1 "github.com/tilt-dev/tilt/pkg/apis/core/v1alpha1"
+	"github.com/tilt-dev/tilt/pkg/logger"
+)
+
+type TiltServerControllerManager struct {
+	config *rest.Config
+	cancel context.CancelFunc
+}
+
+var _ store.SetUpper = &TiltServerControllerManager{}
+var _ store.Subscriber = &TiltServerControllerManager{}
+var _ store.TearDowner = &TiltServerControllerManager{}
+
+func NewTiltServerControllerManager(config *server.APIServerConfig) *TiltServerControllerManager {
+	// TODO(milas): remove this once tests use in-memory connection + real config
+	var cfg *rest.Config
+	if config != nil {
+		cfg = config.GenericConfig.LoopbackClientConfig
+	}
+
+	return &TiltServerControllerManager{
+		config: cfg,
+	}
+}
+
+func (m *TiltServerControllerManager) SetUp(ctx context.Context, st store.RStore) error {
+	if m.config == nil {
+		// TODO(milas): remove this once tests use in-memory connection + real config
+		logger.Get(ctx).Debugf("No REST config provided; controller manager will not be started")
+		return nil
+	}
+
+	scheme := runtime.NewScheme()
+
+	ctx, m.cancel = context.WithCancel(ctx)
+
+	// TODO(milas): we should provide a logr.Logger facade for our logger rather than using zap
+	w := logger.Get(ctx).Writer(logger.DebugLvl)
+	ctrl.SetLogger(zap.New(zap.UseDevMode(true), zap.WriteTo(w)))
+
+	utilruntime.Must(corev1alpha1.AddToScheme(scheme))
+
+	mgr, err := ctrl.NewManager(m.config, ctrl.Options{
+		Scheme:           scheme,
+		Port:             9443,
```

Comment:
this might be a dumb question, but what happens if we set Port to 0? not totally clear why the manager needs a server

Reply:
in particular, will hard-coding a port here break tilt if you try to run two tilt instances at the same time? (this is a common thing people do)

Reply:
I was just looking through that this morning to understand (and see if it's necessary/possible to share the listener with the HUD/apiserver).

It's only used for admission webhooks, and thankfully the webhook server is lazily started only when a webhook is registered with the manager. I'll omit the definition, and we can re-add in the future if we end up using webhooks.

(If we _do_ end up using it, it unfortunately just defaults to `9443` if unspecified/`0`, so we'd need to add options for it.)

Reply:
I changed this to `-1` with an explanation so that in the event someone adds a webhook it'll error out during initialization and we can deal with adding a second port + flag/env vars etc at that point.